IP Userland

hudson-ayers opened the pull request:
### Pull Request Overview

This pull request adds a userland interface to the IP stack. Specifically, it adds an interface via which
userspace applications can send and receive UDP packets. The heavy lifting of this PR is in the userspace driver that enables this, though it also inculdes changes throughout the networing stack that make it possible. This PR is paired with an example application -- IP_Sense -- for sending sensor data from userland using UDP. It is also paired with a udp_rx app that is intended to allow for receiving UDP packets (though this app is currently non-functional).  These two applications are included in a pull request to the tock/libtock-c repository.

### Testing Strategy

Running the userland applications that utilize this interface and verifying that UDP packets are sent and between two boards as expectred. Because the udp_rx application is not yet functional, the ieee802154 radio_rx application was used to verify, as was a packet sniffer.

### TODO or Help Wanted

1. Updates to `imix/src/main.rs` so that the networking stack better utilizes the component interface

2. Fix for the bug that prevents received UDP packets from reaching userspace

3. Greatly improved management of the many addresses and settings that are required to allow for succesful use of the IP stack. A discussion of how this management currently occurs can be found in `doc/ip_config_notes.md` which is part of this PR, along with some of my personal thoughts on how this should change. However, I would greatly appreciate feedback on this topic in particular.

4. Interface beyond just UDP - what about ICMP, or sending raw IP packets?

5. More comprehensive tests that involve varied packet formats and simultaneous reception by multiple applications

### Documentation Updated

- [x] Adds Networking_Userland.md to docs, though this file will still need to be updated again once configuration aspects of the stack are finalized.

- [x] Adds ip_config_notes.md to docs, which includes notes on configuration of the stack

### Formatting

- I am still having issues with my rustfmt version being too new, so I have not run formatall. I will do my best to fix issues if Travis identifies any.

bradjc commented on capsules/src/net/udp/driver.rs:
```diff
@@ -0,0 +1,473 @@
+    pending_tx: Option<[IPAddrPort; 2]>,
+}
+
+impl Default for App {
```
Just derive default on the App struct

bradjc commented on capsules/src/net/udp/driver.rs:
```diff
@@ -0,0 +1,473 @@
+
+                        if let (Some(dst), Some(src)) = (self.parse_ip_port_pair(&cfg.as_ref()[mem::size_of::<IPAddrPort>()..]),
+                                                         self.parse_ip_port_pair(&cfg.as_ref()[..mem::size_of::<IPAddrPort>()])) {
+                            Some([src, dst])
```
What does the app pass as the src address? Is it used anywhere?

hudson-ayers replied:
Right now it is not used anywhere -- it is eventually replaced by the src IP stored in the kernel. I believe at the time this was designed it was unclear whether we would let the src IP addr be chosen by userland apps. I will remove the src address field and add documentation that the
src IP is set in the kernel when I consolidate all of the configuration of the network stack.

bradjc commented on capsules/src/net/udp/driver.rs:
```diff
@@ -0,0 +1,473 @@
+                    }
+                    app.pending_tx = next_tx;
+
+                    self.do_next_tx_sync(appid)
```
I'm not sure how an app would use this interface. `SUCCESS` can either mean "the tx was queued" or "the packet was sent to the radio", and only the former gets a callback when the packet is sent to the radio, correct? How does the app know whether to wait for the callback?

hudson-ayers replied:
I believe a callback is triggered in either case. Regardless of whether the tx was queued or sent immediately, the send_done() callback passed up the radio stack in the kernel is always passed up to userspace via app.tx_callback on line 434ish. @ptcrews is this correct?  

The current implementation (which can be seen in the PR to tock/libtock-c) always waits for this callback.

bradjc replied:
Ok the app is going to wait for the `send_done` callback, and it might get a `packet sent to radio` callback as well, which it likely can just ignore.

I do think it makes sense to add a return value to `SUCCESS` to indicate if the packet was sent to the radio immediately, in case the app wants to use that opportunity to send another packet immediately (rather than having to wait for send done).

hudson-ayers replied:
The app needs to wait for the `packet sent to radio` callback to make sure there was not an error,
if it ignores that callback completely there is a possibility the `send_done` callback never comes, and the app does not know why. I agree that adding a return value to indicate if the packet was immediately sent to the radio is a good idea, implementing that now.

phil-levis replied:
Yeah, there's no such thing as the packet being sent immediately: starting a transmission will require a bunch of asynchronous SPI operations and a failure could occur at any time. Tracking individual packets when there are multiple in flight starts to get tricky, especially if you can't maintain references to them. 

The major reason you'd want to queue as you describe is performance: with an explicit callback, there's a period when the radio is idle, between completion and the callback being issued. For any modern LAN technology this is a real concern, but for low-power link layers, with how fast MCUs, it isn't. But in cases when you do want to be able to enqueue multiple packets to a lower layer you probably want a different trait and intermediate queues.

hudson-ayers replied:
@phil-levis So you are saying that it probably makes more sense for userland apps to always wait for the `send_done` callback anyway before queuing another packet anyway? I buy that, and that means we would not need to change how udp.c is currently implemented in libtock-c. 

However, it still seems that it might be valuable to tell the userland app whether to expect a `packet sent to radio` callback, which could always contain Currently, it seems that there is no way for an app to determine if a packet was sent synchronously or asynchronously, and thus no way for an app to know whether the `packet sent to radio` callback might occur.

hudson-ayers replied:
So how would you recommend we synchronously return an error in the case that the packet is copied into a radio buffer in CPU memory, then sent later once currently in-progress transmissions by other apps finish? In this case the only way I can envision returning these errors would be via a callback, but in the instance that a packet is sent immediately because no other transmissions are in progress it seems much easier to simply directly return any errors that may occur, rather than setting up a callback to pass those errors.

At the end of the day, there are errors which can be generated in the process of passing a packet from the udp driver down to the radio, and if that occurs, we need to inform the user that happened. I agree that the caller should not assume a packet has been sent until it receives a `send_done` callback -- the only purpose of the `packet sent to radio callback` is really to pass back up any errors that might occur before the packet reaches the radio, so that the calling app is not left ignorant to what errors occurred.

Agreed that the different return values of do_next_tx_sync() need to be better documented, I will push some changes with additional documentation within the hour.

phil-levis replied:
I mean, all packets are sent later. Among other things, you usually have to put the radio into a transmit state. I'll take a look at the exact cases for synchronous and asynchronous before commenting more later tonight.

phil-levis replied:
Ah, OK. It looks like the terminology in the comments is a little confusing. The distinction is not about synchronous vs. asynchronous transmission, but rather whether the function signals a callback. The async functions simply call the sync ones, and based on the return value of the sync ones, decide whether to issue a callback (on an error).

More generally, yeah, some errors can be detected synchronously and should be returned immediately. Others might only be discovered asynchronously and so returned in the callback.

bradjc replied:
I agree it would be easier to read the code if those were named differently (as they both use asynchronous operations). Maybe using "immediate" and "queued"?

bradjc commented on capsules/src/net/udp/driver.rs:
```diff
@@ -0,0 +1,473 @@
+                            debug!("Address match!");
+                            let mut app_read = app.app_read.take();
+                            app_read.as_mut().map(|rbuf| {
+                            //app.app_read.take().as_mut().map(|rbuf| {
```
Remove

bradjc commented on capsules/src/net/udp/driver.rs:
```diff
@@ -0,0 +1,473 @@
+    ///
+    /// - `0`: Read buffer. Will contain the received payload.
+    /// - `1`: Write buffer. Contains the UDP payload to be transmitted.
+    /// - `2`: Config buffer. Used to contain miscellaneous data associated with
```
This needs to be documented somewhere.

hudson-ayers replied:
now documented in doc/Networking_Userland.md

bradjc commented on capsules/src/net/udp/driver.rs:
```diff
@@ -0,0 +1,473 @@
+impl<'a> UDPRecvClient for UDPDriver<'a> {
+    fn receive(&self, src_addr: IPAddr, dst_addr: IPAddr, src_port: u16, dst_port: u16, payload: &[u8]) {
+        // debug!("payload: {:?}", payload);
+        debug!("got a payload.");
```
At least comment out the debug!s

alevy commented:
@hudson-ayers FYI, I've not neglected this PR, it's just taking a while to read through

alevy commented on capsules/src/net/udp/driver.rs:
```diff
@@ -0,0 +1,494 @@
+//!
+//! Implements a userspace interface for sending and receiving UDP messages.
+//! Also exposes a list of interface addresses to the application (currently
+//! hard-coded.
```
This comment should describe the high level operation of the driver (e.g. "Processes use this driver to send UDP packets from a common interface and bind to UDP ports for receiving packets." etc).

hudson-ayers replied:
updated

alevy commented on capsules/src/net/udp/driver.rs:
```diff
@@ -0,0 +1,494 @@
+/// Syscall number
+pub const DRIVER_NUM: usize = 0x30002;
+
+const INTERFACES: [IPAddr; 2] = [
```
What are these interfaces?

hudson-ayers replied:
They are intended to represent the interfaces available on the device. There are two for now because that simplified the process of writing a tx/rx test suite - one app uses interface 0, the other uses interface 1. Will be replaced soon with a reference to the source IP address associated with the 15.4 radio interface, once we assign a specific place in the kernel to hold that value.

alevy commented on capsules/src/net/udp/driver.rs:
```diff
@@ -0,0 +1,494 @@
+];
+
+#[derive(Debug)]
+struct IPAddrPort {
```
Maybe `UDPEndpoint` instead? the port is a property of the UDP protocol, not IP

hudson-ayers replied:
Agreed, done.

alevy commented on capsules/src/net/sixlowpan/sixlowpan_state.rs:
```diff
@@ -381,6 +382,11 @@ impl TxState<'a> {
             self.security.set(security);
             self.busy.set(false);
+
+            // ipv6_send ensures passed src pan matches src pan of the underlying 15.4 radio
```
I don't understand this comment. is it meant to be here?

hudson-ayers replied:
It was meant to be here, but it is poorly explained. When I was debugging the networking stack to figure out why the tx/rx userland tests weren't working, one issue I found was that the source pan held in a register on the radio was different from the src pan held in the tx_state struct at the sixlowpan layer. Accordingly, we thought that we had correctly set the PAN of the receiver, but in fact we had only set the PAN used in outgoing messages, and the rf233 was still dropping all messages received from that PAN, as the pan in the radio register was different. To help prevent this, until we clean up the configuration as I discussed on the call, I added the radio_pan as a parameter passed when ipv6_send.rs initializes the 6lowpan tx_state struct.

hudson-ayers replied:
This will all change once the configuration is fixed anyway

hudson-ayers commented on capsules/src/net/udp/driver.rs:
```diff
@@ -0,0 +1,494 @@
+            return None;
+        }
+        let mut pending_app = None;
+        for app in self.apps.iter() {
```
This is correct. I'll document it in the file and in the design doc.

hudson-ayers commented on capsules/src/net/udp/udp_recv.rs:
```diff
@@ -0,0 +1,63 @@
+    fn set_client(&self, client: &'a UDPRecvClient);
+}
+
+pub struct UDPRecvStruct<'a> {
```
This was originally made a trait because it was copied from the IP layer, where it matters that it be a trait because in theory in the future we could have the IP layer sit on top of multiple link layers. I can't imagine we will have UDP on top of any other network layer than IP, so I think it is fine for UDPRecvStruct to become UDPReceiver. I'll do that

hudson-ayers commented:
##Update

I believe I have now addressed all of the original comments on this, other than comments that are related to the configuration of addresses/interfaces for the stack. Please take a look and decide if you think this is ready.

I have also updated the 6lowpan tracking issue to more accurately reflect all of the remaining tasks that are related to 6lowpan.

The most urgent next steps right now include fixing UDP reception to work in userland, and improving configuration of the entire networking stack.

I am happy to work on those and add them as commits to this PR, and delay merging until that is complete. Alternatively, I would be fine with this being merged as-is, and then submitting those updates as an additional PR. Let me know what you guys prefer.

Finally, please take a look at the PR to libtock-c that parallels this PR, when you get the chance. I do not have the same rights on that repo as I do on this repo, so I could not assign any reviewers, but I am sure it needs feedback just as much as this PR did.

hudson-ayers commented:
Userland udp reception is now fixed.

bradjc commented on capsules/src/rf233.rs:
```diff
@@ -912,6 +906,7 @@ impl<S: spi::SpiMasterDevice> spi::SpiMasterClient for RF233<'a, S> {
                 } else {
                     self.state_transition_read(RF233Register::TRX_STATUS, InternalState::READY);
                 }
+                debug!("[RF233] Received packet, sending to client");
```
Presumably don't want to add this debug statement.

bradjc commented on doc/Networking_Userland.md:
```diff
@@ -0,0 +1,266 @@
+
+allow() is used to setup buffers to read/qrite from. This function takes in
+an `allow_num` and a slice. These allow\_nums determine which buffer is being
+setup as follows:
```
Can you move this to a new `.md` file in the syscalls folder to match the other syscall docs?

bradjc commented on doc/Networking_Userland.md:
```diff
@@ -0,0 +1,266 @@
+ - `1`: Write buffer. Contains the UDP payload to be transmitted.
+
+ - `2`: Config buffer. Used to contain miscellaneous data associated with
+        some commands, namely source/destination addresses and ports.
```
This needs to document exactly how that buffer needs to be setup.

bradjc commented on doc/Networking_Userland.md:
```diff
@@ -0,0 +1,266 @@
+- The current tx implementation allows for starvation, e.g. an app with an earlier app ID can
+  starve a later ID by sending constantly.
+
+### POSIX Socket API Functions
```
I'm not sure we need this, but if it is useful can it be added to the networking stack document? Having a userland interface is a part of the full networking stack.

bradjc commented on doc/ip_config_notes.md:
```diff
@@ -0,0 +1,100 @@
+#Userland IP Stack Config
```
Keeping this together with the full networking stack document seems better.

bradjc commented on doc/ip_config_notes.md:
```diff
@@ -0,0 +1,100 @@
+
+##This file contains notes on the configuration of the IP stack.
+
+###Current design (where each of the following values is configuredm and stored):
```
"configured"

niklasad1 commented on boards/imix/src/main.rs:
```diff
@@ -20,6 +20,14 @@ extern crate sam4l;
+use capsules::net::ipv6::ipv6_send::IP6Sender;
+use capsules::net::sixlowpan::{sixlowpan_compression, sixlowpan_state};
+use capsules::net::udp::udp::UDPHeader;
+use capsules::net::udp::udp_recv::{UDPReceiver};
```
remove needless block parentheses on single import!

niklasad1 commented on boards/imix/src/main.rs:
```diff
@@ -133,6 +147,17 @@ static mut RF233_BUF: [u8; radio::MAX_BUF_SIZE] = [0x00; radio::MAX_BUF_SIZE];
+const UDP_HDR_SIZE: usize = 8;
+const PAYLOAD_LEN: usize = 200;
+const DEFAULT_CTX_PREFIX_LEN: u8 = 8;
+static DEFAULT_CTX_PREFIX: [u8; 16] = [0x0 as u8; 16];
```
Needless type-cast and but if you want to be more explicit I think it is more idiomatic to write ```static DEFAULT_CTX_PREFIX: [u8; 16] = [0x0_u8; 16];``` instead!

Also, why `immutable static` instead of `const` here I believe these are equivalent?! (Doesn´t matter to me but would be preferable to be consistent)

niklasad1 commented on boards/imix/src/main.rs:
```diff
@@ -338,6 +447,7 @@ pub unsafe fn reset_handler() {
         ipc: kernel::ipc::IPC::new(),
         ninedof: ninedof,
         radio_driver: radio_driver,
+        udp_driver: udp_driver,
```
it is enough to write **udp_driver ~~:udp driver~~** these days!

niklasad1 commented on capsules/src/net/ipv6/ipv6_recv.rs:
```diff
@@ -0,0 +1,80 @@
+}
+
+pub struct IP6RecvStruct<'a> {
+    client: Cell<Option<&'a IP6RecvClient>>,
```
Use ```OptionalCell``` instead?

niklasad1 commented on capsules/src/net/sixlowpan/sixlowpan_state.rs:
```diff
@@ -312,8 +312,8 @@ pub trait SixlowpanState<'a> {
 /// frames; the `TxState` struct simply produces compressed MAC frames.
 pub struct TxState<'a> {
     // State for the current transmission
```
use ```///``` (rustdoc) for public items!

hudson-ayers commented:
This hasn't been merged yet because bringing it up to date with master results in all userland apps crashing (something to do with PR #1113 causes it). I am not the first to run into this problem, and an issue is forthcoming.

ppannuto commented:
Blocked on #1160

alevy commented:
@hudson-ayers there are conflicts that need to be resolved, i think otherwise it's mergeable.

alevy commented:
bors r+

ppannuto commented:
<img width="791" alt="image" src="https://user-images.githubusercontent.com/339422/45329944-c5ab1900-b517-11e8-9807-29bcf2491c49.png">
Well, I don't know quite what's broken here, but I'm calling this mergable the old way.

notriddle commented:
Looks like Travis built it, but didn't send an update through GitHub's status API?

I have no idea why that would happen, either.
